### PR TITLE
Add platform airdrop functionality to SuperMinterV1_1

### PR DESCRIPTION
**Pull request by Vectorized**
- SuperMinterV1_1:

    - New "PLATFORM_AIRDROP" mode to
    - Removed artist-customizable signer. We don't need that flexibility. Simpler code.

### CODE REVIEW

File: contracts/modules/SuperMinterV1_1.sol [hunk 1]
```diff
@@ -308,15 +335,13 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
...
 
-        // Check if the mint is open.
-        if (LibOps.or(block.timestamp < d.startTime, block.timestamp > d.endTime))
-            revert MintNotOpen(block.timestamp, d.startTime, d.endTime);
-        if (_isPaused(d)) revert MintPaused(); // Check if the mint is not paused.
+        _requireMintOpen(d);
 
         // Perform the sub workflows depending on the mint mode.
         uint8 mode = d.mode;
         if (mode == VERIFY_MERKLE) _verifyMerkle(d, p);
```
Comment (vigneshka) [inline contracts/modules/SuperMinterV1_1.sol:342]: does typescript have switch case with way to error at compile time if a mode isn't handled?

File: contracts/modules/SuperMinterV1_1.sol [hunk 2]
```diff
@@ -395,6 +420,35 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
...
+        unchecked {
+            ISoundEditionV2_1 edition = ISoundEditionV2_1(p.edition);
+
+            uint256 toLength = p.to.length;
+            for (uint256 i; i != toLength; ++i) {
+                edition.mint(p.tier, p.to[i], p.signedQuantity);
+            }
+        }
+
+        emit Presaved(p.edition, p.tier, p.scheduleNum, p.to, p.signedQuantity);
```
Comment (vigneshka) [inline contracts/modules/SuperMinterV1_1.sol:449]: what's the max size of "to" supported before this out of gas?

does omitting the event have any significant gas benefits?

File: contracts/modules/SuperMinterV1_1.sol [hunk 3]
```diff
@@ -100,6 +100,23 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
             ")"
         );
 
+    /**
+     * @dev For EIP-712 presave signature digest calculation.
+     */
+    bytes32 public constant PRESAVE_TYPEHASH =
```
Comment (vigneshka) [inline contracts/modules/SuperMinterV1_1.sol:106]: rename presave and all usages to

platformAirdrop

way more clear what it is

File: contracts/modules/SuperMinterV1_1.sol (hunk 2 above)
Comment (Vectorized) [reply contracts/modules/SuperMinterV1_1.sol:449]: The bulk of gas will be spent on the SSTOREs in the minting. 

My napkin math estimates it at 1000 addresses.

Emitted events do not count towards calldata costs. 30M L2 gas is insignificant compared to calldata costs.

File: contracts/modules/SuperMinterV1_1.sol (hunk 1 above)
Comment (Vectorized) [reply contracts/modules/SuperMinterV1_1.sol:342]: Regular Solidity does not have switch cases.

Comment (vigneshka) [reply contracts/modules/SuperMinterV1_1.sol:342]: 😢

File: contracts/core/SoundEditionV2_1.sol [hunk 4]
```diff
@@ -232,7 +232,7 @@ contract SoundEditionV2_1 is ISoundEditionV2_1, ERC721AQueryableUpgradeable, ERC
         uint8 tier,
         address[] calldata to,
         uint256 quantity
-    ) external payable onlyRolesOrOwner(ADMIN_ROLE) returns (uint256 fromTokenId) {
+    ) external payable onlyRolesOrOwner(ADMIN_ROLE | MINTER_ROLE) returns (uint256 fromTokenId) {
```
Comment (mattg-a16z) [inline contracts/core/SoundEditionV2_1.sol:235]: Checked this in context. The change is to allow `airdrop` to be called by the minter to facilitate the minter's new airdrop functionality. This checks out no response needed

File: contracts/modules/SuperMinterV1_1.sol [hunk 5]
```diff
@@ -303,20 +322,18 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
...
-        if (LibOps.or(block.timestamp < d.startTime, block.timestamp > d.endTime))
-            revert MintNotOpen(block.timestamp, d.startTime, d.endTime);
-        if (_isPaused(d)) revert MintPaused(); // Check if the mint is not paused.
+        _requireMintOpen(d);
 
         // Perform the sub workflows depending on the mint mode.
         uint8 mode = d.mode;
         if (mode == VERIFY_MERKLE) _verifyMerkle(d, p);
         else if (mode == VERIFY_SIGNATURE) _verifyAndClaimSignature(d, p);
+        else if (mode == PLATFORM_AIRDROP) revert InvalidMode();
```
Comment (mattg-a16z) [inline contracts/modules/SuperMinterV1_1.sol:336]: This is important, the main risk creating a new state involves running older functions in unexpected states. This keeps `mintTo` from being called for platform airdrops. This comment is to note down where the expected mitigations are, no response needed.

File: contracts/modules/SuperMinterV1_1.sol [hunk 6]
```diff
@@ -692,6 +719,24 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
             )));
     }
 
+    /**
+     * @inheritdoc ISuperMinterV1_1
+     */
+    function computePlatformAirdropDigest(PlatformAirdrop calldata p) public view returns (bytes32) {
+        // prettier-ignore
+        return
+            _hashTypedData(keccak256(abi.encode(
```
Comment (mattg-a16z) [inline contracts/modules/SuperMinterV1_1.sol:728]: EIP712 digest, addresses most signature gotchas. Good to see.

File: contracts/modules/SuperMinterV1_1.sol [hunk 7]
```diff
@@ -1040,21 +1076,69 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
...
+     * @param p The platform airdrop parameters.
+     */
+    function _verifyAndClaimPlatfromAidropSignature(MintData storage d, PlatformAirdrop calldata p) internal {
+        // Unlike regular signature mints, platform airdrops only use `signedQuantity`.
+        address signer = platformSigner[d.platform];
+        if (!SignatureCheckerLib.isValidSignatureNowCalldata(signer, computePlatformAirdropDigest(p), p.signature))
+            revert InvalidSignature();
+        if (block.timestamp > p.signedDeadline) revert SignatureExpired();
+        uint256 mintId = LibOps.packId(p.edition, p.tier, p.scheduleNum);
+        if (!_claimsBitmaps[mintId].toggle(p.signedClaimTicket)) revert SignatureAlreadyUsed();
```
Comment (mattg-a16z) [inline contracts/modules/SuperMinterV1_1.sol:1139]: So we have the EIP712 signature verification that uses the entire PlatformAirdrop object (except the signature). Then we verify the expiry, then check if it has been used before. Checked toggle to make sure that when the bit is unset it returns true, and that seems to be the case. Reversion should keep toggle from turning the bit off. Should work as intended.

File: contracts/modules/SuperMinterV1_1.sol [hunk 8]
```diff
@@ -285,16 +307,13 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
             d.maxMintable = c.maxMintable;
             d.affiliateFeeBPS = c.affiliateFeeBPS;
             d.mode = c.mode;
-            d.flags = _MINT_CREATED_FLAG | LibOps.toFlag(usePlatformSigner, _USE_PLATFORM_SIGNER_FLAG);
+            d.flags = _MINT_CREATED_FLAG;
```
Comment (mattg-a16z) [inline contracts/modules/SuperMinterV1_1.sol:310]: Sounds like everything is going to use the platform signer. Is that correct?

File: contracts/modules/SuperMinterV1_1.sol [hunk 9]
```diff
@@ -393,6 +410,31 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
...
+     */
+    function platformAirdrop(PlatformAirdrop calldata p) public returns (uint256 fromTokenId) {
+        MintData storage d = _getMintData(LibOps.packId(p.edition, p.tier, p.scheduleNum));
+
+        /* ------------------- CHECKS AND UPDATES ------------------- */
+
+        _requireMintOpen(d);
+
+        if (d.mode != PLATFORM_AIRDROP) revert InvalidMode();
+        _verifyAndClaimPlatfromAidropSignature(d, p);
```
Comment (mattg-a16z) [inline contracts/modules/SuperMinterV1_1.sol:428]: Signature check, expiry check, and replay check are all in this method.

File: contracts/modules/SuperMinterV1_1.sol [hunk 10]
```diff
@@ -393,6 +410,31 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
...
+
+        _incrementPlatformAirdropMinted(d, p);
```
Comment (mattg-a16z) [inline contracts/modules/SuperMinterV1_1.sol:430]: quantity check is inside this method

File: contracts/modules/SuperMinterV1_1.sol [hunk 11]
```diff
@@ -393,6 +410,31 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
...
+
+        /* ------------------------- MINT --------------------------- */
+
+        ISoundEditionV2_1 edition = ISoundEditionV2_1(p.edition);
+        fromTokenId = edition.airdrop(p.tier, p.to, p.signedQuantity);
```
Comment (mattg-a16z) [inline contracts/modules/SuperMinterV1_1.sol:435]: Can only be reached if:
1. the mint is a platform airdrop
2. there is a valid platform signature by the platform specified during mint creation (`createEditionMint` sets the platform that is used to find the signer.
3. the signature hasn't expired
4. the claim ticket number hasn't been used yet
5. the minting doesn't go over the max mintable

As long as this matches the intended restrictions, then this should be good to go.

File: contracts/modules/SuperMinterV1_1.sol [hunk 12]
```diff
@@ -550,25 +596,6 @@ contract SuperMinterV1_1 is ISuperMinterV1_1, EIP712 {
         emit MerkleRootSet(edition, tier, scheduleNum, merkleRoot);
     }
 
-    /**
-     * @inheritdoc ISuperMinterV1_1
-     */
-    function setSigner(
```
Comment (mattg-a16z) [inline contracts/modules/SuperMinterV1_1.sol:556]: This seems to be a change where the only signer used going forward is going to be the platform signer for the platform declared as part of the mint creation. This means no more declaring a specific signer, instead for VERIFY_SIGNATURE mint you'll declare a platform and the platform will will use their enrolled signer. Is this correct? The change on line 1119 seems to support this being the case.

**Review by mattg-a16z — COMMENTED** [review]
I did this in a slightly different way this time. I went through all the key changes to make sure I noted having reviewed the change or asked for verification that my understanding is correct. Overall, it seems like the main changes are:
1. Removing the mint-specific signers in favor of platform signers
2. Addition of another mint type
3. A new minting function (airdrop) to support the new mint type that verifies a platform signature before airdropping to n many addresses
4. Changes to the editions to allow the minter to call the airdrop function directly
5. Changes to configuration functions to keep people from setting price or max mintable for an airdrop mint

If this matches your expectations, then we're probably good. I'll do another once-over tomorrow morning to make sure I'm not missing some strange edge case, but in the mean time, here is the review.

File: contracts/modules/SuperMinterV1_1.sol (hunk 8 above)
Comment (vigneshka) [reply contracts/modules/SuperMinterV1_1.sol:310]: that's correct

File: contracts/modules/SuperMinterV1_1.sol (hunk 12 above)
Comment (vigneshka) [reply contracts/modules/SuperMinterV1_1.sol:556]: correct

**Comment by vigneshka** [comment]
> I did this in a slightly different way this time. I went through all the key changes to make sure I noted having reviewed the change or asked for verification that my understanding is correct. Overall, it seems like the main changes are:
> 
> 1. Removing the mint-specific signers in favor of platform signers
> 2. Addition of another mint type
> 3. A new minting function (airdrop) to support the new mint type that verifies a platform signature before airdropping to n many addresses
> 4. Changes to the editions to allow the minter to call the airdrop function directly
> 5. Changes to configuration functions to keep people from setting price or max mintable for an airdrop mint
> 
> If this matches your expectations, then we're probably good. I'll do another once-over tomorrow morning to make sure I'm not missing some strange edge case, but in the mean time, here is the review.

Perfect summary, thank you Matt!

**Pull request merged**